x-pack/plugin/inference/src/test/java/org/elasticsearch/xpack/inference/rank/textsimilarity/TextSimilarityRankTests.java

```diff
@@ -164,6 +179,27 @@ public void testRerankInferenceResultMismatch() {
         );
     }
 
+    public void testRerankInvalidDocumentIndices() {
+        ElasticsearchAssertions.assertFailures(
+            // Execute search with text similarity reranking
+            client.prepareSearch()
+                .setRankBuilder(
+                    // Simulate reranker returning different number of results from input, also invalid document indices in results
+                    new InvalidInferenceResultCountProvidingTextSimilarityRankBuilder(
+                        "text",
+                        "my-rerank-model",
+                        "my query",
+                        100,
+                        1.5f,
+                        true
+                    )
+                )
+                .setQuery(QueryBuilders.matchAllQuery()),
+            RestStatus.INTERNAL_SERVER_ERROR,
+            containsString("Failed to execute phase [rank-feature], Computing updated ranks for results failed")
```

Comment:
Two things here:

- We should return a 400 response here, the error is ultimately due to a bad value provided by the user
- Can we test that the returned error message contains the root cause of the problem (i.e. "Document and score count mismatch")?

Reply:
The error response is produced [here](https://github.com/elastic/elasticsearch/blob/52834fe041da5cfe4a2b09fcb9b72ba05785a23c/server/src/main/java/org/elasticsearch/action/search/RankFeaturePhase.java#L197), which responds to an `IllegalStateException`. @pmpailis do you know of any way to turn this into a 400?

I'll look into using different matchers in the test to verify the cause.

Reply:
We should work on making the HTTP response code configurable. 500 == rest suppressed errors == things to investigate during serverless on-call

Reply:
Update: there's no method in [ElasticsearchAssertions#assertFailures](https://github.com/elastic/elasticsearch/blob/52834fe041da5cfe4a2b09fcb9b72ba05785a23c/test/framework/src/main/java/org/elasticsearch/test/hamcrest/ElasticsearchAssertions.java#L496) that we can use to check the cause of a shard failure 🙁.

Reply:
I think that we could slightly rewrite this to capture the cause's message (as there won't be any response to decref). E.g:
```
        SearchPhaseExecutionException ex = expectThrows(
            SearchPhaseExecutionException.class,
            client.prepareSearch()
                .setRankBuilder(
                    // Simulate reranker returning different number of results from input, also invalid document indices in results
                    new InvalidInferenceResultCountProvidingTextSimilarityRankBuilder(
                        "text",
                        "my-rerank-model",
                        "my query",
                        100,
                        1.5f,
                        true
                    )
                )
                .setQuery(QueryBuilders.matchAllQuery())::get
        );
        assertThat(ex.getDetailedMessage(), containsString("Document and score count mismatch"));
```

Reply:
I think that we could slightly rewrite this to capture the cause's message (as there won't be any response to decref).
E.g:
```
        SearchPhaseExecutionException ex = expectThrows(
            SearchPhaseExecutionException.class,
            client.prepareSearch()
                .setRankBuilder(
                    // Simulate reranker returning different number of results from input, also invalid document indices in results
                    new InvalidInferenceResultCountProvidingTextSimilarityRankBuilder(
                        "text",
                        "my-rerank-model",
                        "my query",
                        100,
                        1.5f,
                        true
                    )
                )
                .setQuery(QueryBuilders.matchAllQuery())::get
        );
        assertThat(ex.getDetailedMessage(), containsString("Document and score count mismatch"));
```

Reply:
> @pmpailis do you know of any way to turn this into a 400?

The status of the error is determined through the `SearchPhaseExecutionException#status`, which in turn uses the cause for the exception through `org.elasticsearch.ExceptionsHelper#status` .  I think converting the `IllegalStateException` to an `IllegalArgumentException` should be enough to have a 400 response.